mindmeld/models/embedder_models.py

```diff
@@ -0,0 +1,184 @@
+            os.makedirs(folder)
+
+        if not os.path.exists(self.cache_path) or os.path.getsize(self.cache_path) == 0:
+            self.cache = {}
+        else:
+            with open(self.cache_path, "rb") as fp:
+                self.cache = pickle.load(fp)
+
+        self.model = self.load(**kwargs)
+
+    @abstractmethod
+    def load(self, **kwargs):
+        """Loads the embedder model
+
+        Returns:
+            The model object.
+        """
+        pass
+
+    @abstractmethod
+    def encode(self, text_list):
+        """
+        Args:
+            text_list (list): A list of text strings for which to generate the embeddings.
+
+        Returns:
+            (list): A list of numpy arrays of the embeddings.
+        """
+        pass
+
+    def clear_cache(self):
+        """Deletes the cache file.
+        """
+        if os.path.exists(self.cache_path):
+            os.remove(self.cache_path)
+
+    def get_encodings(self, text_list):
+        """Fetches the encoded values from the cache, or generates them.
+        Args:
+            text_list (list): A list of text strings for which to get the embeddings.
+
+        Returns:
+            (list): A list of numpy arrays with the embeddings.
+        """
+        encoded = [self.cache.get(text, None) for text in text_list]
+        cache_miss_indices = [i for i, vec in enumerate(encoded) if vec is None]
+        text_to_encode = [text_list[i] for i in cache_miss_indices]
+        model_encoded_text = self.encode(text_to_encode)
+
+        for i, v in enumerate(cache_miss_indices):
+            encoded[v] = model_encoded_text[i]
+            self.cache[text_to_encode[i]] = model_encoded_text[i]
+        return encoded
+
+    def dump(self):
+        """Dumps the cache to disk.
+        """
+        with open(self.cache_path, "wb") as fp:
+            pickle.dump(self.cache, fp)
+
+
+class BertEmbedder(Embedder):
+    """
+    Encoder class for bert models as described here: https://github.com/UKPLab/sentence-transformers
+    """
+
+    def load(self, **kwargs):
+        DEFAULT_BERT = "bert-base-nli-mean-tokens"
+        if self.model_name == "default":
+            bert_model_name = DEFAULT_BERT
+            logger.info("No bert model specifications passed, using default.")
+        else:
+            bert_model_name = self.model_name
+        return SentenceTransformer(bert_model_name)
+
+    def encode(self, text_list):
+        return self.model.encode(text_list)
+
+
+def get_query_tokens(query):
+    """Splits into tokens, removes punctuation, and removes whitespace.
+    """
+    query = query.translate(str.maketrans("", "", string.punctuation))
+    tokens = query.split()
+    tokens = [t.strip() for t in tokens]
+    return tokens
+
+
+class GloveEmbedder(Embedder):
+    """
+    Encoder class for GloVe embeddings as described here: https://nlp.stanford.edu/projects/glove/
+    """
+
+    def load(self, **kwargs):
+        DEFAULT_EMBEDDING_DIM = 300
```

Comment:
Put this as a class constant so we can access the default embedding dim

Reply:
Why would you want to access this outside of the load method? These classes are not exposed to app developers to use directly

Reply:
I think in the case if the developer wants to ask "what is the default embedding dim" it is easier to look up class constants instead of inspecting the code.

Reply:
Not sure I understand, how would it be easier? They would still have to inspect the code, just one line above where it is now, right? If the developer is looking for defaults they would most likely check the documentation?

Reply:
They can use auto-hinting (`GloveEmbedder. DEFAULT_EMBEDDING_DIM`) or look at the generated API reference. For example, you can see the constant `CLF_TYPE` here:
https://www.mindmeld.com/docs/apidoc/mindmeld.components.role_classifier.html#mindmeld.components.role_classifier.RoleClassifier.CLF_TYPE